src/engine/PDKind.cc

```diff
@@ -0,0 +1,295 @@
+            f_abs_conj = logic.mkAnd(f_abs_conj, versionedFla);
+
+            }
+
+        PTRef not_fabs = logic.mkNot(lemma.lemma);
+        PTRef versioned_not_fabs = tm.sendFlaThroughTime(not_fabs, currentUnrolling);
+        PTRef t_k_constr = logic.mkAnd(t_k, f_abs_conj);
+
+        MainSolver solver1(logic, config, "PDKIND");
+        solver1.insertFormula(iframe_abs);
+        solver1.insertFormula(t_k_constr);
+        solver1.insertFormula(versioned_not_fabs);
+        auto res1 = solver1.check();
+
+
+        if (res1 == s_False) {
+            newIframe.insert(lemma);
+            continue;
+        }
+
+        auto model1 = solver1.getModel();
+
+        PTRef fcex = tm.sendFlaThroughTime(lemma.counter_example, currentUnrolling);
+        
+        MainSolver solver2(logic, config, "PDKKIND");
+        solver2.insertFormula(iframe_abs);
+        solver2.insertFormula(t_k_constr);
+        solver2.insertFormula(fcex);
+        auto res2 = solver2.check();
+        
+        if (res2 == s_True) {
+            auto model2 = solver2.getModel();
+            PTRef g_cex = reachability_checker.generalize(*model2, t_k, fcex);
+            // u kazdeho cex si pamatovat pocet korku
+            // g_cex je fcex.pocetkroku + k
+            auto reach_res = reachability_checker.checkReachability(n - k + 1, n, g_cex);
+            if (std::get<0>(reach_res)) {
+                // g_cex pocet kroku + pocet kroku v reachibility checkingu
+                invalid = true;
+                continue;
+            } else {
+                PTRef g_abs = std::get<2>(reach_res);
+                IFrameElement newLemma = IFrameElement(g_abs, g_cex);
+                /*
+                 *
+                 *
+                 * */
+                iframe.insert(newLemma);
+                q.push(newLemma);
+                q.push(lemma);
+                continue;
+            }
+        }
+
+        PTRef g_cti = reachability_checker.generalize(*model1, t_k, versioned_not_fabs);
+        auto reach_res = reachability_checker.checkReachability(n - k + 1, n, g_cti);
+        if (std::get<0>(reach_res)) {
+            reach_res = reachability_checker.checkReachability(n + 1, std::get<1>(reach_res) + k, not_fabs);
+            np = std::min(np, std::get<1>(reach_res));
+            IFrameElement newLemma = IFrameElement(logic.mkNot(lemma.counter_example), lemma.counter_example);
+            iframe.insert(newLemma);
+            newIframe.insert(newLemma);
+        } else {
+            PTRef g_abs = std::get<2>(reach_res);
+            g_abs = logic.mkAnd(lemma.lemma, g_abs);
+            IFrameElement newLemma = IFrameElement(g_abs, lemma.counter_example);
+            iframe.insert(newLemma);
+            iframe.extract(lemma);
+            q.push(newLemma);
+        }
+    }
+    PushResult res(iframe, newIframe, np, invalid);
+    return res;
+}
+
+std::tuple<bool, PTRef> ReachabilityChecker::reachable (int k, PTRef formula) {
```

Comment:
```suggestion
std::tuple<bool, PTRef> ReachabilityChecker::reachable(int k, PTRef formula) {
```

Also, maybe `k` should be `unsigned`?